source/common/filesystem/win32/filesystem_impl.cc

```diff
@@ -145,6 +149,34 @@ std::string InstanceImplWin32::fileReadToEnd(const std::string& path) {
   return file_string.str();
 }
 
+std::string InstanceImplWin32::fileReadToEnd(const std::string& path,
+                                             bool retry_sharing_violations) {
+  std::string out;
+  int attempts = 1;
+  constexpr int max_attempts = 3;
+  do {
+    try {
+      out = fileReadToEnd(path);
+      return out;
+    } catch (const EnvoyException& e) {
+      if (retry_sharing_violations) {
+        std::string exception_message(e.what());
+        ENVOY_LOG_MISC(warn, "{}", exception_message);
+        if (exception_message.find("Error sharing violation") != std::string::npos) {
```

Comment:
I think your point about the caller owning how the file access is handled is a valid one. I would maybe implement this using `Create/OpenFile` (so we can check error codes directly rather than exception strings) in a retry loop in the original `fileReadToEnd` or another abstraction in `FilesystemSubscriptionImpl` before we attempt to read the file.